lib/fhir/query-utils.js

```diff
@@ -450,40 +459,85 @@ module.exports = (mongo) => {
-      // fetch the referenced filters using reference FHIR module
-      chainedModule.searchFilters(ctx, (err, badRequest, filters) => {
-        if (err) { return callback(err) }
-        if (badRequest) { return callback(new Error(`Bad request made to ${chainedModule.name} module while chaining parameters`)) }
-        mongo.getDB((err, db) => {
+      const fetchReferencedFilters = (chainedParam, chainedModule, callback) => {
+        const ctx = { query: {} }
+        ctx.query[chainedParam] = value
+
+        // fetch the referenced filters using reference FHIR module
+        chainedModule.searchFilters(ctx, (err, badRequest, filters) => {
           if (err) { return callback(err) }
-          const c = db.collection(chainedModule.name)
-          // find patients in question
-          c.find(filters).project({ id: 1 }).toArray((err, results) => {
+          if (badRequest) { return callback(new Error(`Bad request made to ${chainedModule.name} module while chaining parameters`)) }
+          mongo.getDB((err, db) => {
             if (err) { return callback(err) }
-            if (results.length === 0) {
-              // return query to a non-existant resource so an empty result set is returned
-              const query = {}
-              query[baseParam] = `${chainedModule.name}/_none`
-              return callback(null, query)
-            }
-
-            // build up a query that finds docs for each matching resource
-            const clauses = []
-            results.forEach((resource) => {
-              const clause = {}
-              clause[baseParam] = `${chainedModule.name}/${resource.id}`
-              clauses.push(clause)
+            const c = db.collection(chainedModule.name)
+            // find patients in question
+            c.find(filters).project({ id: 1 }).toArray((err, results) => {
+              if (err) { return callback(err) }
+              if (results.length === 0) {
+                // return query to a non-existant resource so an empty result set is returned
+                const query = {}
+                query[baseParam] = `${chainedModule.name}/_none`
+                return callback(null, query)
+              }
+
+              // build up a query that finds docs for each matching resource
+              const clauses = []
+              results.forEach((resource) => {
+                const clause = {}
+                clause[baseParam] = `${chainedModule.name}/${resource.id}`
+                clauses.push(clause)
+              })
+              if (clauses.length > 1) {
+                callback(null, { '$or': clauses })
+              } else {
+                callback(null, clauses[0])
+              }
             })
-            if (clauses.length > 1) {
-              callback(null, { '$or': clauses })
-            } else {
-              callback(null, clauses[0])
-            }
           })
         })
-      })
+      }
+
+      if (Array.isArray(chainedModule)) {
```

Comment:
A cleaner way to do this might be to make it an array of 1 element if it isn't an array. Then we can get rid of the big if blocks.